From cf56abd471fecc60fc760278b2767bf83e2483fd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Sun, 10 Mar 2024 15:48:26 +0100
Subject: [PATCH] Use built-in isoformat support

The ISO formatted string can be stored in multiple formats. Use the built-in datetime.isoformat() and datetime.fromisoformat() functions, which able to handle all variants.

Fixes: https://github.com/libkeepass/pykeepass/issues/382
---
 pykeepass/pykeepass.py | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/pykeepass/pykeepass.py b/pykeepass/pykeepass.py
index 08be860..57b301c 100644
--- a/pykeepass/pykeepass.py
+++ b/pykeepass/pykeepass.py
@@ -28,7 +28,6 @@
 BLANK_DATABASE_FILENAME = "blank_database.kdbx"
 BLANK_DATABASE_LOCATION = os.path.join(os.path.dirname(os.path.realpath(__file__)), BLANK_DATABASE_FILENAME)
 BLANK_DATABASE_PASSWORD = "password"
-DT_ISOFORMAT = "%Y-%m-%dT%H:%M:%S%fZ"
 
 class PyKeePass():
     """Open a KeePass database
@@ -804,7 +803,7 @@ def _encode_time(self, value):
                 struct.pack('<Q', diff_seconds)
             ).decode('utf-8')
         else:
-            return value.strftime(DT_ISOFORMAT)
+            return value.isoformat()
 
     def _decode_time(self, text):
         """datetime.datetime: Convert base64 time or plaintext time to datetime"""
@@ -819,9 +818,9 @@ def _decode_time(self, text):
                     )
                 )
             except BinasciiError:
-                return datetime.strptime(text, DT_ISOFORMAT).replace(tzinfo=timezone.utc)
+                return datetime.fromisoformat(text).replace(tzinfo=timezone.utc)
         else:
-            return datetime.strptime(text, DT_ISOFORMAT).replace(tzinfo=timezone.utc)
+            return datetime.fromisoformat(text).replace(tzinfo=timezone.utc)
 
 def create_database(
         filename, password=None, keyfile=None, transformed_key=None
